org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/ASTParser.java

```diff
@@ -1362,6 +1338,7 @@ private ASTNode internalCreateASTCached(IProgressMonitor monitor) {
 	 * @see ASTNode#getLength()
 	 */
 	private ASTNode internalCreateASTForKind() {
+		// TODO make it independent from ECJ
```

Comment:
no TODOS in master please

Reply:
I can just remove the TODO for the convention, but that would be kind of lying: there would still be things to do here, things that won't be done very soon. So how to track such "will need more work" spots in JDT?
To be honest, I think a convention against TODOs is kind of missing the idea that software development can be incremental.

Reply:
The problem with TODOS in JDT Core is that they tend to silently stay forever - currently already 260 matches :-(
Until the TODOs are solved its better to stay on a branch, like the compiler guys do it with new Java features.

Reply:
> The problem with TODOS in JDT Core is that they tend to silently stay forever

That's not only in JDT, it's the nature of TODOs.
TODOs rarely become highest priority -they're not at the time they're written, otherwise they would be fixed. Yet they're valuable hints for people who come to this code as they point to hotspots where to look at. I've seen some TODOs here and there actually describing a problem that was not so important for a while, and suddenly important enough to consider, and the TODO pointing to the right location and action.

But I'm curious here, is the "no TODO" rule a rule that is backed by the whole project or just your opinion?
ie would it be an objective blocker for a future merge, or are we just discussing it "casually" because why not?

> Until the TODOs are solved its better to stay on a branch, like the compiler guys do it with new Java features.

I strongly disagree with that. This approach of keeping upcoming critical value on a branch is much harder for new contributor to join development and testing. It's IMO a bad practice, and -without willing to be disrespectful of the legacy workflows that were relevant at some point- they're also 1 reason why JDT is facing difficulties coping with new Java releases.
I'd rather avoid replicating this bad practice.

Reply:
Regarding TODOs,  it would appear almost all of the TODO listed here are in what is a new bundle. This seems very reasonable to me, as newer bundles, whose logic only gets enabled by proactively setting some new setting, are usually under a fast-paced development and are not expected to have all of their implementation functional immediately. 

If @mickaelistria  had peppered TODOs throughout the existing jdt.core code, I would agree that this is not good practice and could leave the codebase in a messy and incomplete state. 

But in this case, I disagree.